File: remote-agents/orchestrator/src/lib/session.js
// session.js — per-run session folder + transcript writer.
//
// Each call to startSession() creates remote-agents/sessions/<iso>-<slug>/
// with a prompt.txt and an empty transcript.jsonl. The returned session
// handle has helpers for appending events and finalizing.

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const SESSIONS_ROOT = path.resolve(__dirname, '../../sessions');

function slugify(s) {
  return (s || '')
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-|-$/g, '')
    .slice(0, 40) || 'session';
}

function tsForFilename() {
  // 2026-05-28T13-11-42-491Z — safe on Windows (no colons)
  return new Date().toISOString().replace(/[:.]/g, '-');
}

export function startSession({ projectDir, projectName, userPrompt, flowName }) {
  fs.mkdirSync(SESSIONS_ROOT, { recursive: true });

  const id = `${tsForFilename()}-${slugify(flowName || projectName || 'run')}`;
  const dir = path.join(SESSIONS_ROOT, id);
  fs.mkdirSync(dir);

  const promptFile = path.join(dir, 'prompt.txt');
  const transcriptFile = path.join(dir, 'transcript.jsonl');
  const metaFile = path.join(dir, 'meta.json');

  fs.writeFileSync(promptFile, userPrompt || '', 'utf8');
  fs.writeFileSync(transcriptFile, '');
  fs.writeFileSync(metaFile, JSON.stringify({
    id,
    flowName: flowName ?? null,
    projectName: projectName ?? null,
    projectDir: projectDir ?? null,
    userPrompt: userPrompt ?? null,
    startedAt: new Date().toISOString(),
  }, null, 2));

  return {
    id,
    dir,
    promptFile,
    transcriptFile,
    metaFile,
    flowName,
    projectName,
    projectDir,
    userPrompt,
    startedAt: Date.now(),
  };
}

export function appendEvent(session, event) {
  const line = JSON.stringify({ t: new Date().toISOString(), ...event }) + '\n';
  fs.appendFileSync(session.transcriptFile, line);
}

export function endSession(session, finalEvent = {}) {
  appendEvent(session, { kind: 'end', ...finalEvent });
  const meta = JSON.parse(fs.readFileSync(session.metaFile, 'utf8'));
  meta.endedAt = new Date().toISOString();
  meta.durationMs = Date.now() - session.startedAt;
  meta.result = finalEvent.result ?? null;
  fs.writeFileSync(session.metaFile, JSON.stringify(meta, null, 2));
  return meta;
}

File: remote-agents-dotnet/src/RemoteAgents/Core/Sessions/Session.cs
using System.Diagnostics;
using System.Text.Json;
using System.Text.RegularExpressions;
using RemoteAgents.Primitives;

namespace RemoteAgents.Sessions;

public sealed record StartSessionRequest(
    string? ProjectDir,
    string? ProjectName,
    string? UserPrompt,
    string? FlowName);

// A per-run session folder under remote-agents-dotnet/sessions/<id>/.
// Owns three files: prompt.txt, meta.json, transcript.jsonl. Holds no
// running state beyond paths; sinks write transcript entries, the agent
// lifecycle owns meta finalization.
public sealed class Session
{
    private static readonly Regex SlugStrip = new("[^a-z0-9]+", RegexOptions.Compiled);
    private readonly Stopwatch _sw;

    public string Id { get; }
    public string Dir { get; }
    public string PromptFile { get; }
    public string TranscriptFile { get; }
    public string MetaFile { get; }
    public SessionMeta Meta { get; private set; }

    private Session(string id, string dir, SessionMeta meta)
    {
        Id = id;
        Dir = dir;
        PromptFile = Path.Combine(dir, "prompt.txt");
        TranscriptFile = Path.Combine(dir, "transcript.jsonl");
        MetaFile = Path.Combine(dir, "meta.json");
        Meta = meta;
        _sw = Stopwatch.StartNew();
    }

    public static Session Start(StartSessionRequest req, string? sessionsRoot = null)
    {
        sessionsRoot ??= DefaultSessionsRoot();
        Directory.CreateDirectory(sessionsRoot);

        var slug = Slugify(req.FlowName ?? req.ProjectName ?? "run");
        var id = $"{TsForFilename()}-{slug}";
        var dir = Path.Combine(sessionsRoot, id);
        Directory.CreateDirectory(dir);

        File.WriteAllText(Path.Combine(dir, "prompt.txt"), req.UserPrompt ?? "");
        File.WriteAllText(Path.Combine(dir, "transcript.jsonl"), "");

        var meta = new SessionMeta
        {
            Id = id,
            Orchestrator = "csharp",
            SchemaVersion = "1",
            FlowName = req.FlowName,
            ProjectName = req.ProjectName,
            ProjectDir = req.ProjectDir,
            UserPrompt = req.UserPrompt,
            StartedAt = DateTimeOffset.UtcNow,
        };

        var s = new Session(id, dir, meta);
        s.WriteMeta();
        return s;
    }

    public void End(SessionResult result, string? failureReason = null)
    {
        Meta.EndedAt = DateTimeOffset.UtcNow;
        Meta.DurationMs = _sw.ElapsedMilliseconds;
        Meta.Result = result;
        Meta.FailureReason = failureReason;
        WriteMeta();
    }

    // Typed artifact accessors. Centralizes the basename → file mapping
    // that used to be sprinkled across flow files. Callers reference the
    // enum value; the basename only appears here.
    public string GetArtifactPath(SessionArtifact artifact) =>
        Path.Combine(Dir, ArtifactBasename(artifact));

    // Static overload for callers that have a sessionDir string but not a
    // Session instance (e.g. Reviews helpers that take sessionDir as a
    // param). Same basename mapping — the literal filename still appears
    // exactly once.
    public static string GetArtifactPath(string sessionDir, SessionArtifact artifact) =>
        Path.Combine(sessionDir, ArtifactBasename(artifact));

    public Task WriteArtifactAsync(SessionArtifact artifact, string contents, CancellationToken ct = default) =>
        File.WriteAllTextAsync(GetArtifactPath(artifact), contents, ct);

    public Task<string?> ReadArtifactAsync(SessionArtifact artifact, CancellationToken ct = default)
    {
        var path = GetArtifactPath(artifact);
        return File.Exists(path) ? File.ReadAllTextAsync(path, ct).ContinueWith(t => (string?)t.Result, ct)
                                 : Task.FromResult<string?>(null);
    }

    private static string ArtifactBasename(SessionArtifact artifact) => artifact switch
    {
        SessionArtifact.Transcript    => "transcript.jsonl",
        SessionArtifact.ClaudeText    => "claude-text.txt",
        SessionArtifact.ClaudeRaw     => "claude-raw.txt",
        SessionArtifact.CodexReview   => "codex-review.txt",
        SessionArtifact.CodexReviewJl => "codex-review.jsonl",
        _ => throw new ArgumentOutOfRangeException(nameof(artifact)),
    };

    private void WriteMeta()
    {
        var json = JsonSerializer.Serialize(Meta, SessionJsonContext.Default.SessionMeta);
        File.WriteAllText(MetaFile, json);
    }

    private static string Slugify(string s)
    {
        var lower = s.ToLowerInvariant();
        var stripped = SlugStrip.Replace(lower, "-").Trim('-');
        if (stripped.Length > 40) stripped = stripped[..40];
        return stripped.Length == 0 ? "session" : stripped;
    }

    // 2026-05-28T13-11-42-491Z — safe on Windows (no colons)
    private static string TsForFilename() =>
        DateTimeOffset.UtcNow.ToString("yyyy-MM-ddTHH-mm-ss-fffZ");

    private static string DefaultSessionsRoot() =>
        OrchestratorPaths.SessionsRoot()
        ?? Path.Combine(Environment.CurrentDirectory, "sessions");
}
